File: src/lib/src/main/java/work/lclpnet/ap2/impl/util/world/SizedSpaceFinder.java
package work.lclpnet.ap2.impl.util.world;

import net.minecraft.core.BlockPos;
import net.minecraft.world.entity.EntityDimensions;
import net.minecraft.world.entity.EntityType;
import net.minecraft.world.level.BlockGetter;
import net.minecraft.world.phys.Vec3;
import net.minecraft.world.phys.shapes.BooleanOp;
import net.minecraft.world.phys.shapes.Shapes;
import net.minecraft.world.phys.shapes.VoxelShape;
import org.jetbrains.annotations.Nullable;
import work.lclpnet.ap2.api.util.world.SpaceFinder;

import java.util.Iterator;
import java.util.List;
import java.util.Objects;
import java.util.Spliterators;
import java.util.stream.StreamSupport;

public class SizedSpaceFinder implements SpaceFinder {

    private final BlockGetter blockView;
    private final float halfWidth, height, halfLength;

    public SizedSpaceFinder(BlockGetter blockView, float width, float height, float length) {
        this.blockView = blockView;
        this.halfWidth = width * 0.5f;
        this.height = height;
        this.halfLength = length * 0.5f;
    }

    @Override
    public List<Vec3> findSpaces(Iterator<BlockPos> positions) {
        var spliterator = Spliterators.spliteratorUnknownSize(positions, 0);

        return StreamSupport.stream(spliterator, false)
                .map(this::spaceAt)
                .filter(Objects::nonNull)
                .toList();
    }

    @Nullable
    private Vec3 spaceAt(BlockPos pos) {
        final double minX = pos.getX(), minY = pos.getY(), minZ = pos.getZ();
        final double maxX = minX + 1, maxY = minY + 1, maxZ = minZ + 1;

        // prefer the block center, then fall back to the edges
        double[] xs = {minX + 0.5, minX, maxX};
        double[] zs = {minZ + 0.5, minZ, maxZ};

        for (double x : xs)
            for (double z : zs)
                for (double y = minY; y <= maxY; y += 0.5)
                    if (hasSpace(x, y, z)) return new Vec3(x, y, z);

        return null;
    }

    private boolean hasSpace(double x, double y, double z) {
        double minX = x - halfWidth, minZ = z - halfLength;
        double maxX = x + halfWidth, maxY = y + height, maxZ = z + halfLength;

        VoxelShape space = Shapes.create(minX, y, minZ, maxX, maxY, maxZ);

        return BlockPos.betweenClosedStream(
                (int) Math.floor(minX),
                (int) Math.floor(y),
                (int) Math.floor(minZ),
                (int) Math.ceil(maxX),
                (int) Math.ceil(maxY),
                (int) Math.ceil(maxZ)
        ).noneMatch(pos -> {
            VoxelShape shape = blockView.getBlockState(pos).getCollisionShape(blockView, pos)
                    .move(pos.getX(), pos.getY(), pos.getZ());

            return Shapes.joinIsNotEmpty(shape, space, BooleanOp.AND);
        });
    }

    public static SizedSpaceFinder create(BlockGetter blockView, EntityType<?> entityType) {
        EntityDimensions dimensions = entityType.getDimensions();

        int width = (int) Math.ceil(dimensions.width());
        int height = (int) Math.ceil(dimensions.height());

        return new SizedSpaceFinder(blockView, width, height, width);
    }
}

File: src/lib/src/main/kotlin/work/lclpnet/ap2/util/world/SizedSpaceFinder.kt
package work.lclpnet.ap2.util.world

import net.minecraft.core.BlockPos
import net.minecraft.world.entity.EntityType
import net.minecraft.world.level.BlockGetter
import net.minecraft.world.phys.Vec3
import net.minecraft.world.phys.shapes.BooleanOp
import net.minecraft.world.phys.shapes.Shapes
import work.lclpnet.ap2.api.util.world.SpaceFinder
import java.util.*
import java.util.stream.StreamSupport
import kotlin.math.ceil
import kotlin.math.floor

class SizedSpaceFinder(
    private val blockView: BlockGetter,
    private val width: Float,
    private val height: Float,
    private val length: Float
) : SpaceFinder {

    override fun findSpaces(positions: Iterator<BlockPos>): List<Vec3> {

        val spliterator = Spliterators.spliteratorUnknownSize(positions, 0)

        @Suppress("UNCHECKED_CAST")
        return StreamSupport.stream(spliterator, false)
            .map { pos -> spaceAt(pos) }
            .filter { obj -> Objects.nonNull(obj) }
            .toList() as List<Vec3>
    }

    private fun spaceAt(pos: BlockPos): Vec3? {
        val minX = pos.x.toDouble()
        val minY = pos.y.toDouble()
        val minZ = pos.z.toDouble()
        val maxX = minX + 1
        val maxY = minY + 1
        val maxZ = minZ + 1

        // prefer the block center, then fall back to the edges
        val xs = doubleArrayOf(minX + 0.5, minX, maxX)
        val zs = doubleArrayOf(minZ + 0.5, minZ, maxZ)

        for (x in xs) for (z in zs) {
            var y = minY
            while (y <= maxY) {
                if (hasSpace(x, y, z)) return Vec3(x, y, z)
                y += 0.5
            }
        }

        return null
    }

    private fun hasSpace(x: Double, y: Double, z: Double): Boolean =
        hasSpaceAt(blockView, x, y, z, width.toDouble(), height.toDouble(), length.toDouble())

    companion object {

        @JvmStatic
        fun create(blockView: BlockGetter, entityType: EntityType<*>): SizedSpaceFinder {
            val dimensions = entityType.dimensions

            val width = ceil(dimensions.width().toDouble()).toInt()
            val height = ceil(dimensions.height().toDouble()).toInt()

            return SizedSpaceFinder(blockView, width.toFloat(), height.toFloat(), width.toFloat())
        }

        fun hasSpaceAt(
            level: BlockGetter,
            x: Double,
            y: Double,
            z: Double,
            width: Double,
            height: Double,
            length: Double
        ): Boolean {
            val halfWidth = width * 0.5
            val halfLength = length * 0.5

            val minX = x - halfWidth
            val minZ = z - halfLength
            val maxX = x + halfWidth
            val maxY = y + height
            val maxZ = z + halfLength

            val space = Shapes.create(minX, y, minZ, maxX, maxY, maxZ)

            return BlockPos.betweenClosedStream(
                floor(minX).toInt(),
                floor(y).toInt(),
                floor(minZ).toInt(),
                ceil(maxX).toInt(),
                ceil(maxY).toInt(),
                ceil(maxZ).toInt()
            ).noneMatch { pos ->
                val shape = level.getBlockState(pos).getCollisionShape(level, pos)
                    .move(pos.x.toDouble(), pos.y.toDouble(), pos.z.toDouble())

                Shapes.joinIsNotEmpty(shape, space, BooleanOp.AND)
            }
        }
    }
}